SE-1166

rajeeshknambiar opened the pull request:
SE-1166

Add waiting/preparing/processing start/end/total time and the total number of records and processed number of records to the EDI Document.

armandcela commented on addons/edi/models/edi_document.py:
```diff
@@ -260,6 +261,52 @@ class EdiDocument(models.Model):
     fail_fast = fields.Boolean(related="doc_type_id.fail_fast", readonly=True)
 
+    # Processing statistics
+    waiting_time_start = fields.Datetime(string="Waiting start", readonly=True)
```
Is it thought to be a one2many with edi.document example edi.document.statistics which have fields, 
start,end,total,processed phase(waiting,preparing and executing maybe)

It seems like are similar fields added all to edi.document currently

rajeeshknambiar replied:
Implemented in d2d0059987663e679e903383986dc3a068e681fa

armandcela replied:
Sorry i think my fault, hasnt explained well what i meant :(.

armandcela replied:
Resolving , but i still believe is better to be in one2many

PeterAlabaster replied:
Discussed this w/ Armand, I think we should change it to be relational (one2many) as this gives us traceability of previous state transitions and wait times, for instance - if a record is prepared, moved back to draft, and re-prepared.

The only downside with this approach is we would need to also show this new model in the UI so that we can search across all states and times. In which case, we'd probably also need a `document_type_id` on the `edi.processing.stats` model so they can search by document type.

I think we should remove the `related=` time fields from the `edi.document` too

PeterAlabaster replied:
Relevant discussion:
RKV:
```
Another question about SE-1166 (EDI processing times and no. of records, https://unipart.atlassian.net/browse/SE-1166) to the POs (just had a call with Armand where he shared experience on Collingwood):

1. The total records and total records processed — should those be updated realtime (similar to the progress log), while the document is still executing? And here the "total records processed" means currently executed bunch?

2. The timing information are now added to EDI document as in the attached screenshot: this would allow for filtering documents based on these fields (Miquel clarified in the notes that it is required to "run reports on them"); OR is it okay to show these timing details in a list view (similar to say, sale order lines; which is a bit more tricky when filtering edi documents based on these fields)?
```

AP:
```
1. Let's leave the real time updates, just once the doc has been processed is fine.

2. Having the metrics on the form is fine. We can still search and report on them from there.
```

adamrpatrick commented on addons/edi/models/edi_document.py:
```diff
@@ -310,6 +357,42 @@ def _get_state_name(self):
+        for rec in self:
+            if rec.waiting_time_start:
+                diff = (rec.waiting_time_end or fields.Datetime.now()) - rec.waiting_time_start
+                rec.waiting_time_total = diff.total_seconds() / 3600.0
```
It might be better to have these total times in seconds rather than hours.

rajeeshknambiar replied:
Unfortunately, `float_time` widget requires the value in hours (see https://www.odoo.com/documentation/14.0/developer/reference/javascript/javascript_reference.html#non-relational-fields)

adamrpatrick commented on addons/edi/models/edi_document.py:
```diff
@@ -260,6 +261,52 @@ class EdiDocument(models.Model):
+    # Processing statistics
+    waiting_time_start = fields.Datetime(string="Waiting start", readonly=True)
+    waiting_time_end   = fields.Datetime(string="Waiting end", readonly=True)
+    waiting_time_total = fields.Float(
```
Add help string and specify the time units the total is calculated in.

rajeeshknambiar replied:
Added in d2d0059987663e679e903383986dc3a068e681fa

adamrpatrick replied:
I still don't see this?

armandcela commented on addons/edi/models/edi_document.py:
```diff
@@ -260,6 +261,52 @@ class EdiDocument(models.Model):
+        readonly=True,
+        help="Total processing time since receiving",
+    )
+    total_records = fields.Integer(
```
Question: What is the difference between total_records and total_records_processed?
Question2: Are these fields committed while is being executed example?

adamrpatrick replied:
It might be an idea just to log the total_records_processed once the doc has executed and remove the total_records field.

armandcela replied:
If only the state is the difference, i think dont need 2 fields. We can understand that are total processed if edi doc is in done state

armandcela replied:
Do we need to save the information we log on server logs, to be visible from the users while is being executed so the know they are executed these records out of total records. Practically similar information with server logs. maybe it is and didnt understand which part of code does this

rajeeshknambiar replied:
`total_records_processed` is the number of successfully processed records.

armandcela replied:
total_records is the number of total records i suppose :). If EDI is done they show the same information and the first field will be 0 till is done isnt it?

rajeeshknambiar replied:
For document types set as not fail fast; it might be possible that the total records and total processed records can differ, though (if some records fail, for instance).

adamrpatrick commented on addons/edi_sale/models/edi_sale_line_request_record.py:
```diff
@@ -17,6 +17,15 @@ class EdiDocument(models.Model):
+    def _compute_edi_rec_count(self):
+        if self.sale_line_request_ids:
+            for doc in self:
+                doc.total_records = len(doc.sale_line_request_ids)
```
Wouldn't this conflict with the totals from other related fields on the EDI document? E.G. edi_sale_request_records.

rajeeshknambiar replied:
This one is tricky. There are some specialized classes with reference fields to both header & item records where we calculated the total of both; but for some classes like the `edi_sale` the header (`edi.sale.request.record`) & item (`edi.sale.line.request.record`) records are in different classes. It seems we can have only one of those; unless we find a clever way...

adamrpatrick replied:
In this case, let's go with sale lines seeing as this will be the larger of the two.

rajeeshknambiar replied:
Just to clarify: the sale lines method will get called first and since it doesn't propagate `super()` call, the sale request method wouldn't get called anyway. So we may leave it as is.

adamrpatrick commented on addons/edi_product/models/edi_product_record.py:
```diff
@@ -13,6 +13,15 @@ class EdiDocument(models.Model):
         "edi.inactive.product.record", "doc_id", string="Inactive Products"
     )
 
+    @api.depends("product_ids", "state")
```
How often does this method get called e.g. when processing an item master? One call for each row on the input CSV? It could get expensive for large CSVs.

rajeeshknambiar replied:
Did some profiling, it gets called 4 times altogether when processing a file (one each before & after prepare & execute; no matter how large). Should be good from performance point of view.

(Specifically checked udes_itmstr).

PeterAlabaster commented on addons/edi/models/edi_document.py:
```diff
@@ -699,3 +780,85 @@ class EdiDocumentProgress(models.Model):
+
+class EdiDocumentStats(models.Model):
+    """EDI Document processing statistics such as start/end of waiting/prepare/execute, total time
+    for each processing stage, total records and total processed records etc."""
```
We don't seem to check total records or total processed records on this model, like the docstring says

rajeeshknambiar replied:
Fixed in d1ad09b

PeterAlabaster commented on addons/edi/models/edi_document.py:
```diff
@@ -260,6 +261,39 @@ class EdiDocument(models.Model):
+        readonly=True,
+        help="Total number of associated edi.record",
+    )
+    total_records_processed = fields.Integer(
```
Does this update as the EDI is being processed? i.e if we have a really large file, does this field reflect the _current_ state of the execution, or does it only update at the end?

rajeeshknambiar replied:
For the moment, it is not updated during processing, only at the end.

PeterAlabaster replied:
POs wanted this to be updated as the edi is being processed, not at the end

PeterAlabaster commented on addons/edi/models/edi_document.py:
```diff
@@ -310,13 +344,32 @@ def _get_state_name(self):
         vals = dict(self.fields_get(allfields=["state"])["state"]["selection"])
         return vals[self.state]
 
+    def _compute_edi_rec_count(self):
```
I'm really not sure this should be computed.

This also smells a bit to me as it is something that has to be implemented on each EDI (as we can see by the amount of changes in these MRs), and it's another thing to remember to do each time we add a new EDI.

I think the intention with this was to basically do the same thing the logging is doing (i.e ` _logger.info("Executed %s in %.2fs, %d queries", self.name, stats.elapsed, stats.count)`) and then it only needs to be in one place. That is already handled by the way EDI records are batched in groups of 1,000's. That is the only level of granularity we require, as there's not much point in having lower granularity due to the speed in which rows are processed..

So, can't we just replace all of these `_compute_edi_rec_count` hooks to just one single hook just before/after that logger entry? There's also a similar one for preparing, in `public/odoo_edi/addons/edi/models/edi_document.py::action_execute() and ::action_prepare()`

rajeeshknambiar replied:
That would be a good change (but the same information is already in the progress log, though). I'll defer to @adamrpatrick based on previous discussions.

adamrpatrick replied:
I think it's worth exploring if we can achieve more or less the intention of the story without having to implement a new computed method for each document type.

Commit:
[ADD] edi: compute total records and total processed records and processing times for an EDI document.

SE-1166

Signed-off-by: Rajeesh KV <rajeesh.kv@unipart.com>

PeterAlabaster commented on addons/edi/models/edi_document.py:
```diff
@@ -390,10 +423,36 @@ def commit_processing_state_change(self, processing_state=""):
+        now = fields.Datetime.now()
+        
+        # Stop the current state's processing time
+        current_state = self.processing_state or "waiting"
```
Q: Did we want them to be able to have multiple lines of the same state for traceability? Maybe it's not needed.. It's certainly simpler this way

rajeeshknambiar replied:
The DO for this one is `Start, end and total time of processing states: waiting, preparing, executing`, so I reckon having multiple lines for same state may not be required.

Commit:
[ADD] edi: new menu, tree/form/search view for edi.document.stats.

This will help with reporting/filtering edi.document based on the document name/type etc.

SE-1166

Signed-off-by: Rajeesh KV <rajeesh.kv@unipart.com>

armandcela commented on addons/edi/models/edi_document.py:
```diff
@@ -12,6 +12,11 @@
 
 AutodetectDocument = namedtuple("AutodetectDocument", ["type", "inputs"])
 
+PROCESSING_STATES = [
```
Nit: Indentation doesnt look right, maybe there are tabs and spaces mixed

rajeeshknambiar replied:
Updated.

armandcela commented on addons/edi/models/edi_document.py:
```diff
@@ -390,10 +423,36 @@ def commit_processing_state_change(self, processing_state=""):
+        EdiDocumentStats = self.env["edi.document.stats"]
+
+        self.ensure_one()
+        assert(self.processing_state != processing_state)
```
Q: Is this assert needed?

rajeeshknambiar replied:
I'd rather keep it, in case.

PeterAlabaster commented on addons/edi/models/edi_document.py:
```diff
@@ -260,6 +266,12 @@ class EdiDocument(models.Model):
 
+    # Processing timing statistics
+    doc_stat_ids = fields.One2many(
+        "edi.document.stats", "doc_id", string="Document Statistics", readonly=True, index=True
```
nit: `ondelete="cascade"` to prevent any leftover `edi.document.stats` records if an EDI document is deleted

PeterAlabaster replied:
Or it might be the other way round, in which case - you've already done it :laughing:
